File: apps/worker/src/career_worker/providers/ports.py
from __future__ import annotations

from dataclasses import dataclass
from typing import Any, Protocol, TypeVar

T = TypeVar("T")


@dataclass(frozen=True, slots=True)
class GenerationRequest:
    model: str
    prompt_hash: str
    input: Any
    output_schema_version: str
    max_tokens: int | None = None


@dataclass(frozen=True, slots=True)
class GenerationResult:
    output: Any
    provider: str
    model: str
    model_version: str
    input_hash: str
    schema_version: str


class GenerationPort(Protocol):
    async def generate(self, request: GenerationRequest) -> GenerationResult: ...


class EmbeddingPort(Protocol):
    async def embed(self, text: str, model: str) -> tuple[list[float], int, str, str]: ...


class RerankerPort(Protocol):
    async def rerank(self, query: str, candidates: list[str]) -> list[tuple[int, float]]: ...

File: packages/ai/src/ports.ts
export interface GenerationRequest {
  model: string;
  promptHash: string;
  input: unknown;
  outputSchemaVersion: string;
  maxTokens?: number;
}

export interface GenerationResult<T> {
  output: T;
  provider: string;
  model: string;
  modelVersion: string;
  inputHash: string;
  schemaVersion: string;
}

export interface GenerationPort {
  generate<T>(request: GenerationRequest): Promise<GenerationResult<T>>;
}
export interface EmbeddingPort {
  embed(
    input: string,
    model: string
  ): Promise<{ vector: number[]; dimensions: number; modelVersion: string; inputHash: string }>;
}
export interface RerankerPort {
  rerank(
    query: string,
    candidates: readonly string[]
  ): Promise<readonly { index: number; score: number }[]>;
}
export interface ToolProposal {
  name: string;
  arguments: Record<string, unknown>;
  requiresConfirmation: true;
}
